File: services/gameClock.ts
/**
 * GAME CLOCK ARCHITECTURE (Prompt 01 of 20: FOUNDATION)
 *
 * Primary temporal engine for Destiny.
 *
 * ARCHITECTURAL MANDATE:
 * Defaults to YEARLY ticks for macro life progression, but is structurally
 * designed to accept variable resolution (MONTHLY, WEEKLY, DAILY) for active events
 * (courtroom trials, business turnarounds, medical crises, military tours)
 * coming in Prompts 02-20.
 */

import { ClockState, TickResolution } from '../types';

export const RESOLUTION_SUBTICKS: Record<TickResolution, number> = {
  YEARLY: 1,
  MONTHLY: 12,
  WEEKLY: 52,
  DAILY: 365,
};

export class GameClockService {
  private state: ClockState;

  constructor(initialYear = 2000, initialResolution: TickResolution = 'YEARLY') {
    this.state = {
      year: initialYear,
      subTick: 1,
      resolution: initialResolution,
      isZoomedIn: initialResolution !== 'YEARLY',
      totalTicksElapsed: 0,
    };
  }

  public getState(): ClockState {
    return { ...this.state };
  }

  /**
   * Advance simulation time by [steps] in current resolution.
   */
  public tick(steps = 1): ClockState {
    if (steps <= 0) throw new Error('Tick steps must be greater than zero.');

    let { year, subTick, resolution, totalTicksElapsed } = this.state;
    const maxSubTicks = RESOLUTION_SUBTICKS[resolution];

    for (let i = 0; i < steps; i++) {
      totalTicksElapsed++;
      if (resolution === 'YEARLY') {
        year++;
      } else {
        subTick++;
        if (subTick > maxSubTicks) {
          subTick = 1;
          year++;
        }
      }
    }

    this.state = {
      ...this.state,
      year,
      subTick,
      totalTicksElapsed,
    };

    return this.getState();
  }

  /**
   * Enters higher-resolution temporal zoom for active interactive events.
   * Architectural hook for subsequent prompts.
   */
  public enterEventZoom(resolution: Exclude<TickResolution, 'YEARLY'>): ClockState {
    this.state = {
      ...this.state,
      resolution,
      subTick: 1,
      isZoomedIn: true,
    };
    return this.getState();
  }

  /**
   * Exits zoom and returns to standard yearly macro-simulation.
   */
  public exitEventZoom(): ClockState {
    this.state = {
      ...this.state,
      resolution: 'YEARLY',
      subTick: 1,
      isZoomedIn: false,
    };
    return this.getState();
  }

  public formatDisplay(): string {
    const { year, subTick, resolution } = this.state;
    switch (resolution) {
      case 'YEARLY':
        return `Year ${year}`;
      case 'MONTHLY':
        return `Year ${year} • Month ${subTick}/12`;
      case 'WEEKLY':
        return `Year ${year} • Week ${subTick}/52`;
      case 'DAILY':
        return `Year ${year} • Day ${subTick}/365`;
    }
  }
}

File: android/app/src/main/java/com/example/services/GameClock.kt
package com.example.services

/**
 * GAME CLOCK ARCHITECTURE (Prompt 01 of 20 Foundation)
 *
 * Core timekeeper for the Destiny simulation.
 * Defaults to YEARLY resolution, but is specifically architected to support
 * variable resolution (MONTHLY, WEEKLY, DAILY) for zoom-in active events
 * (e.g., military deployment, political campaign, courtroom trials, surgical emergencies)
 * in subsequent prompts.
 */
class GameClock(
    initialYear: Int = 2000,
    initialResolution: TickResolution = TickResolution.YEARLY
) {
    enum class TickResolution(val subTicksPerYear: Int, val label: String) {
        YEARLY(1, "Annual"),
        MONTHLY(12, "Monthly"),
        WEEKLY(52, "Weekly"),
        DAILY(365, "Daily")
    }

    data class ClockState(
        val year: Int,
        val subTick: Int, // 1-indexed relative to resolution (e.g., Month 1..12, Week 1..52)
        val resolution: TickResolution,
        val isZoomedIn: Boolean = false,
        val totalTicksElapsed: Long = 0L
    )

    private var state: ClockState = ClockState(
        year = initialYear,
        subTick = 1,
        resolution = initialResolution,
        isZoomedIn = initialResolution != TickResolution.YEARLY
    )

    fun getCurrentState(): ClockState = state

    /**
     * Advance the clock by [steps] of current resolution.
     */
    fun advanceTick(steps: Int = 1): ClockState {
        require(steps > 0) { "Steps must be positive" }
        var currentYear = state.year
        var currentSubTick = state.subTick
        val maxSubTicks = state.resolution.subTicksPerYear
        var elapsed = state.totalTicksElapsed

        repeat(steps) {
            elapsed++
            if (state.resolution == TickResolution.YEARLY) {
                currentYear++
            } else {
                currentSubTick++
                if (currentSubTick > maxSubTicks) {
                    currentSubTick = 1
                    currentYear++
                }
            }
        }

        state = state.copy(
            year = currentYear,
            subTick = currentSubTick,
            totalTicksElapsed = elapsed
        )
        return state
    }

    /**
     * Zoom into higher resolution (e.g., monthly/weekly) during an active encounter.
     * Architectural hook for future prompts (Prompts 02-20).
     */
    fun enterEventZoom(temporaryResolution: TickResolution) {
        state = state.copy(
            resolution = temporaryResolution,
            subTick = 1,
            isZoomedIn = true
        )
    }

    /**
     * Revert to standard yearly simulation once an active event completes.
     */
    fun exitEventZoom() {
        state = state.copy(
            resolution = TickResolution.YEARLY,
            subTick = 1,
            isZoomedIn = false
        )
    }

    fun formatDisplay(): String {
        return when (state.resolution) {
            TickResolution.YEARLY -> "Year ${state.year}"
            TickResolution.MONTHLY -> "Year ${state.year} • Month ${state.subTick}/12"
            TickResolution.WEEKLY -> "Year ${state.year} • Week ${state.subTick}/52"
            TickResolution.DAILY -> "Year ${state.year} • Day ${state.subTick}/365"
        }
    }
}
